### PR TITLE
[Feature][MRV2] Support basic MLA prefill context parallelism

### PR DESCRIPTION
**Pull request by wzx0726**
### What this PR does / why we need it?

Add basic MRV2 prefill context parallelism support for MLA attention on Ascend.

This PR:

- propagates per-request prefill state through MRV2 attention metadata
- adds a shared PCP configuration helper
- adds the MLA PCP metadata, metadata builder, and attention implementation
- gathers rank-local prefill KV data and writes the complete KV cache
- preserves one-token prefill classification under PCP/DCP
- adds unit coverage for backend selection, slot mappings, KV gathering, padding, and short prefill classification

ACL Graph and prefix caching support are intentionally excluded and will be submitted separately.

The small `enable_pcp` helper is intentionally included in both this PR and #14023 so the MLA and GQA PRs can be reviewed and validated independently. After either PR merges, the remaining PR will be rebased and the duplicate helper diff will be dropped.

### Does this PR introduce _any_ user-facing change?

Yes. MLA models can use MRV2 prefill context parallelism on Ascend for the supported eager/PIECEWISE execution paths.

### How was this patch tested?

- Added focused unit tests for common prefill metadata propagation
- Added MLA PCP metadata and KV gather unit tests
- Added a regression test for one-token prefill classification
- `git diff --check upstream/main...HEAD`
- AST syntax validation for all changed Python files

The focused pytest cases were not executed locally because pytest is not installed in the current Windows environment. The PR is being opened as a draft for review and CI validation.

- vLLM version: v0.27.1
- vLLM main: https://github.com/vllm-project/vllm/commit/cdc4824a21eaa986d4d1fee90a7e6465c9f706e6

### CODE REVIEW

File: vllm_ascend/attention/mla_v1.py [hunk 1]
```diff
@@ -1772,3 +1875,94 @@ def forward(
...
+
+        cos = attn_metadata.prefill.cos
+        sin = attn_metadata.prefill.sin
+        prefill_q_pe = self.rope_single(prefill_q_pe, cos, sin)
+
+        local_prefill_kv = kv_no_split[num_decode_tokens:local_num_input_tokens]
+        padded_cos = self._pad_tokens(cos, local_prefill_capacity)
+        padded_sin = self._pad_tokens(sin, local_prefill_capacity)
+
+        pcp_group = get_pcp_group()
```
Comment (pisceskkk) [inline vllm_ascend/attention/mla_v1.py:1919]:
```
    def exec_kv_prefill(...):
        local_prefill_kv = kv_no_split[num_decode_tokens:local_num_input_tokens]
        padded_cos = self._pad_tokens(cos, local_prefill_capacity)
        padded_sin = self._pad_tokens(sin, local_prefill_capacity)
        pcp_group = get_pcp_group()
        rank_slot_mappings = attn_metadata.pcp_slot_mapping.view(
            pcp_group.world_size,
            local_num_input_tokens,
        )
        expanded_prefill_slots = rank_slot_mappings[:, num_decode_tokens:].flatten()
        (gathered_kv, gathered_cos, gathered_sin), gathered_prefill_slots = _gather_prefill_cache_inputs(
            (local_prefill_kv, padded_cos, padded_sin),
            expanded_prefill_slots,
            num_decode_tokens=0,
        )
        gathered_k_pe, gathered_k_c_normed = super().exec_kv_prefill(
            gathered_kv,
            gathered_cos,
            gathered_sin,
            kv_cache,
            gathered_prefill_slots,
        )
        local_prefill_start = pcp_group.rank_in_group * local_prefill_capacity
        local_prefill_end = local_prefill_start + local_num_prefill_tokens
        prefill_k_pe = gathered_k_pe[local_prefill_start:local_prefill_end]
        prefill_k_c_normed = gathered_k_c_normed[local_prefill_start:local_prefill_end]
        return prefill_k_pe, prefill_k_c_normed

```

File: vllm_ascend/attention/mla_v1.py [hunk 2]
```diff
@@ -243,6 +266,15 @@ def __init__(
...
         scheduler_config = vllm_config.scheduler_config
+        parallel_config = vllm_config.parallel_config
+
+        pcp_enabled = parallel_config.prefill_context_parallel_size > 1
+        dcp_enabled = parallel_config.decode_context_parallel_size > 1
+
+        # The regular MLA path may classify short extensions as decodes. PCP
+        # and DCP must instead preserve the scheduler's per-request prefill
+        # classification.
+        self.treat_short_extends_as_decodes = not (pcp_enabled or dcp_enabled)
```
Comment (weiguihua2) [inline vllm_ascend/attention/mla_v1.py]: There is no need to use self.

File: vllm_ascend/attention/mla_v1.py [hunk 3]
```diff
@@ -670,6 +702,77 @@ def build_for_graph_capture(
...
+        metadata.slot_mapping = torch.cat(
+            (
+                rank_slot_mappings[0, :num_decode_tokens],
+                rank_slot_mappings[
+                    self.pcp_rank,
+                    num_decode_tokens : metadata.num_actual_tokens,
+                ],
+            )
+        )
+        metadata.pcp_slot_mapping = expanded_slot_mapping
```
Comment (weiguihua2) [inline vllm_ascend/attention/mla_v1.py]: No need for pcp_slot_mapping, use slot_mapping directly.

File: vllm_ascend/attention/mla_v1.py [hunk 4]
```diff
@@ -1772,3 +1884,72 @@ def forward(
...
+        local_prefill_capacity = local_num_input_tokens - num_decode_tokens
+        local_num_prefill_tokens = num_actual_tokens - num_decode_tokens
+        if kv_no_split.shape[0] != local_num_prefill_tokens:
+            raise RuntimeError(
+                f"PCP MLA prefill input length mismatch: {kv_no_split.shape[0]} != {local_num_prefill_tokens}."
+            )
+
+        local_prefill_kv = self._pad_tokens(kv_no_split, local_prefill_capacity)
+        padded_cos = self._pad_tokens(cos, local_prefill_capacity)
+        padded_sin = self._pad_tokens(sin, local_prefill_capacity)
```
Comment (wzx0726) [inline vllm_ascend/attention/mla_v1.py]: Check whether the upstream of vllm is padded and whether the padding is deleted.

File: vllm_ascend/worker/v2/attn_utils.py [hunk 5]
```diff
@@ -115,6 +100,7 @@ def build_attn_metadata(
     slot_mappings: torch.Tensor,
     kv_cache_config: KVCacheConfig,
     dcp_local_seq_lens: torch.Tensor | None = None,
+    is_prefilling: torch.Tensor | None = None,
```
Comment (wzx0726) [inline vllm_ascend/worker/v2/attn_utils.py]: Confirm whether prefilling is needed and whether it is within this PR.

File: tests/ut/attention/a2/test_mla_cp.py [hunk 6]
```diff
@@ -18,10 +18,205 @@
...
     AscendMLAMetadataBuilder,
+    AscendMLAPCPImpl,
+    AscendMLAPCPMetadata,
+    AscendMLAPCPMetadataBuilder,
     AscendMLAPrefillMetadata,
 )
 
 
+def _make_pcp_metadata(
+    *,
```
Comment (wzx0726) [inline tests/ut/attention/a2/test_mla_cp.py]: test代码

File: tests/ut/attention/a2/test_mla_cp.py [hunk 7]
```diff
@@ -18,10 +18,205 @@
...
     AscendMLAMetadata,
     AscendMLAMetadataBuilder,
+    AscendMLAPCPImpl,
+    AscendMLAPCPMetadata,
+    AscendMLAPCPMetadataBuilder,
     AscendMLAPrefillMetadata,
 )
 
 
+def _make_pcp_metadata(
```
Comment (wzx0726) [inline tests/ut/attention/a2/test_mla_cp.py]: test检视

File: vllm_ascend/attention/mla_v1.py [hunk 8]
```diff
@@ -1772,3 +1884,72 @@ def forward(
...
+            gathered_kv,
+            gathered_cos,
+            gathered_sin,
+            kv_cache,
+            gathered_prefill_slots,
+        )
+        local_prefill_start = pcp_group.rank_in_group * local_prefill_capacity
+        local_prefill_end = local_prefill_start + local_num_prefill_tokens
+        prefill_k_pe = gathered_k_pe[local_prefill_start:local_prefill_end]
+        prefill_k_c_normed = gathered_k_c_normed[local_prefill_start:local_prefill_end]
```
Comment (wzx0726) [inline vllm_ascend/attention/mla_v1.py]: keep local value before splitting

File: vllm_ascend/attention/mla_v1.py
Comment (pisceskkk) [inline vllm_ascend/attention/mla_v1.py]: Could we extract this unpad logic from the MLA forward into a separate function, and then override that function in the PCP implementation to skip the unpad step?

File: vllm_ascend/worker/v2/attn_utils.py (hunk 5 above)
Comment (pisceskkk) [reply vllm_ascend/worker/v2/attn_utils.py]: please confirm here

**Comment by imsatoshi** [comment]
I prepared a focused follow-up for the remaining validation/test gaps: https://github.com/wzx0726/vllm-ascend/pull/1 (commit `c45d74563`). It removes the private vLLM PCP helper dependency, restores fail-fast spec/full-graph/sparse-graph checks, fixes cached backend-selection tests, and adds a two-card exact-token PCP parity gate. Eight focused tests plus Ruff/diff checks pass locally; the NPU E2E is added but was not run because the available shared hosts were occupied.

**Comment by imsatoshi** [comment]
Update: the validation/fix follow-up is now in the upstream repository as #14316 (the temporary contributor-fork MR was closed). Real Ascend 910B PCP=2 testing reproduced a vLLM v0.27.1 base-manager initialization failure; #14316 fixes the release compatibility path, and the rerun achieved exact greedy-token parity against PCP=1 for three uneven/chunked prompts (24/24 output tokens).

File: vllm_ascend/attention/mla_v1.py [hunk 9]
```diff
@@ -1772,3 +1891,76 @@ def forward(
...
+class AscendMLAPCPImpl(AscendMLAImpl):
+    """MRV2 MLA implementation for Prefill Context Parallelism."""
+
+    def _get_num_prefill_kv_tokens(
+        self,
+        attn_metadata: AscendMLAMetadata,
+    ) -> int:
+        if not isinstance(attn_metadata, AscendMLAPCPMetadata):
+            raise RuntimeError("PCP MLA prefill requires AscendMLAPCPMetadata.")
+        return (
```
Comment (pisceskkk) [inline vllm_ascend/attention/mla_v1.py]: nit: maybe we should add a comment to clarify why we use `pcp_local_num_input_tokens` here (such as for pcp pad, and where we unpad)

**Review by pisceskkk — APPROVED** [review]
LGTM, leave a nit.

**Review by weiguihua2 — DISMISSED** [review]

**Review by weijinqian0 — DISMISSED** [review]

**Comment by wzx0726** [comment]
/rerun
[Bot]: rerun completed.

**Failed:**
- E2E: still in progress, retry /rerun after completion

**Comment by wzx0726** [comment]
/rerun
[Bot]: rerun completed.

**Failed:**
- E2E: some jobs could not be re-run (still waiting on dependencies)

**Comment by zhangxinyuehfad** [comment]
/rerun
[Bot]: rerun completed.

**Failed:**
- E2E: some jobs could not be re-run (still waiting on dependencies)

**Review by weijinqian0 — APPROVED** [review]

**Comment by wzx0726** [comment]
/rerun
[Bot]: rerun completed.

**Failed:**
- E2E: some jobs could not be re-run (still waiting on dependencies)

**Comment by wzx0726** [comment]
/rerun

[Bot]: rerun completed.

**Rerun (failed jobs only):**
- E2E

**Comment by wzx0726** [comment]
/rerun
[Bot]: rerun completed.

**Rerun (failed jobs only):**
- E2E
